move resource options from Config to hostConfig

hqhq opened the pull request:
Cgroup resources are host dependent, they should be in hostConfig.

For backward compatibility, we just copy it to hostConfig, and leave it in
Config for now, so there is no regressions, but the right way to use this
throught json is to put it in HostConfig, like:
  {
      "Hostname": "",
      ...
      "HostConfig": {
          "Memory": 314572800,
      ...
      }
  }

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

hqhq commented:
@crosbymichael @unclejack @LK4D4 please take a look at this one, once this is good to go, the movement for other options will on their way soon.

Changes for test cases will in subsequent PR, these changes won't made any test cases fail.

LK4D4 commented:
@hqhq cool, thank you! But I think that this will be in 1.18 api version and we're now in sort of "code freeze"
@crosbymichael What should we do?

crosbymichael commented:
approach looks good to me.  Maybe we should do all the fields in one PR so we get everything updated in one merge.

SvenDowideit commented:
and something in the release notes and API recent changes for those that were using the API.

hqhq commented:
updated, add all resource option changes in this PR.

@SvenDowideit do you mean add something for `docs/sources/release-notes.md`?
Shouldn't this be changed when we really release a version? Or do I need to do something in this PR?

SvenDowideit commented:
I prefer to keep all related updates in one PR, that way they hit the repo together, and can't be forgotten. but tbh, I'm not sure what the process is.

so this PR is only about moving and renaming, and then #9536 adds better help?

in that case, Docs LGTM @fredlf @jamtur01

jamtur01 commented:
Docs LGTM

hqhq commented:
ping @crosbymichael @LK4D4

unclejack commented:
@hqhq You've got to rebase the PR because it can't be merged.

hqhq commented:
@unclejack @crosbymichael @LK4D4 
Rebased.

fredlf commented on docs/sources/reference/api/docker_remote_api_v1.17.md:
```diff
@@ -180,7 +180,8 @@ Json Parameters:
 -   **CpuShares** - An integer value containing the CPU Shares for container
       (ie. the relative weight vs othercontainers).
-    **CpuSet** - String value containg the cgroups Cpuset to use.
+-   **Cpuset** - The same as CpusetCpus, but obsoleted, please don't use.
```
obsoleted --> deprecated

hqhq replied:
Done.

fredlf commented:
A couple of copy-edits, otherwise docs LGTM.

SvenDowideit commented:
I presume these API changes need to move to the 1.18 API doc

cpuguy83 commented:
Why are we changing the CLI flag?

jfrazelle commented:
ping @ashahab-altiscale for the lxc template change

ashahab-altiscale commented:
@jfrazelle Thanks for letting me know. Can you start an lxc build on this?

hqhq commented:
@cpuguy83 Do you mean we change --cpuset to --cpuset-cpus?
Because it's a better name, cpuset is not only for cpuset.cpus, but also for cpuset.mems, I was about to add support for cpuset.mems, by then, the name --cpuset would not be properly.

hqhq commented:
Rebased and move the API change to v1.18 API doc.

cpuguy83 commented:
@hqhq ah, ok.  Also didn't notice originally that you did actually deprecate the old syntax properly.
So this LGTM.

Probably want to go ahead and squash commits.

vbatts commented:
LGTM

jfrazelle commented:
LGTM just needs commits to be squashed :) and maybe @crosbymichael because he is very good at catching weird things in the configs

hqhq commented:
Hi, sorry for the late, I'm not sure how these commits can be squashed any more, are you proposing to make all the changes in one commit?
I think it's better to do one thing in one commit, if you think it's better to take the whole movement as one thing, it's ok for me to squash it, just let me know what you think :)

hqhq commented:
Hi, this PR is depended by a lot of PRs, can we pick this up so we can move on?
@crosbymichael

hqhq commented:
ping.
@crosbymichael @jfrazelle @LK4D4

hqhq committed:
move resources from Config to HostConfig

Cgroup resources are host dependent, they should be in hostConfig.

For backward compatibility, we just copy it to hostConfig, and leave it in
Config for now, so there is no regressions, but the right way to use this
throught json is to put it in HostConfig, like:
  {
      "Hostname": "",
      ...
      "HostConfig": {
	  "CpuShares": 512,
          "Memory": 314572800,
          ...
      }
  }

As we will add CpusetMems, CpusetCpus is definitely a better name, but some
users are already using Cpuset in their http APIs, we also make it compatible.

The main idea is keep using Cpuset in Config Struct, and make it has the same
value as CpusetCpus, but not always, some scenarios:
 - Users use --cpuset in docker command, it can setup cpuset.cpus and can
   get Cpuset field from docker inspect or other http API which will get
   config info.
 - Users use --cpuset-cpus in docker command, ditto.
 - Users use Cpuset field in their http APIs, ditto.
 - Users use CpusetCpus field in their http APIs, they won't get Cpuset field
   in Config info, because by then, they should already know what happens
   to Cpuset.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq commented:
@crosbymichael @jfrazelle @LK4D4 @cpuguy83 
Rebased again, I squash all commits into one commit, hope that's what you want.

Please review this, it have been taken too long (more than 4 months since the original PR), a lot of cgroup improvement are depending on this one, hope we can make some progress soon. Thanks.

hqhq commented:
ping @LK4D4 @crosbymichael @jfrazelle

icecrime commented:
As far as I can tell, this PR has both code and docs LGTM, and the commits have been squashed as requested. Thanks for you patience @hqhq, I'm merging this hoping that nobody will have any objections.
